### ISSUE TITLE
[Ascend] fused topk_softmax produces nondeterministic corrupted weights when renormalize=True

### ISSUE THREAD

**Issue by luentong** [FlagOpen/FlagGems#5123]
## Summary

On the Ascend NPU backend, `flag_gems.fused.topk_softmax` returns corrupted and nondeterministic `topk_weights` when `renormalize=True`.

The selected expert indices remain correct and deterministic. The issue affects router weights only.

## Environment

- Device: Ascend910B4-1
- CANN: 9.0.0
- Python: 3.11
- See attached `environment.txt` for exact package versions.

## Minimal Reproduction

Input shape reproducing the issue:

- logits: `(64, 256)`, `torch.bfloat16`, NPU
- topk: `8`
- `renormalize=True`

The attached script runs the same input repeatedly and compares against:

```python
torch.softmax(logits.float(), dim=-1)
torch.topk(...)
```

## Observed Behavior Before Fix

With renormalize=False:
top-k weights match the FP32 PyTorch reference
max error: approximately 7.45e-09
20 repeated executions are deterministic
expert ids are correct

With renormalize=True:
expert ids remain correct and deterministic
top-k weights become corrupted and nondeterministic
observed maximum values reach approximately 3e9
row sums reach approximately 1e10
repeated executions differ by approximately 1e9

This is not normal floating-point error.

## Suspected Cause
The kernel first stores raw top-k weights to global memory, then, in the same kernel, loads those output values back from global memory to apply renormalization.
On the Ascend Triton backend, this global-memory read-after-write pattern appears nondeterministic.

## Proposed Fix
Keep the first kernel responsible only for raw softmax/top-k output. When renormalize=True, launch a second kernel on the same stream to normalize the already-written top-k weights.
This removes the same-kernel global-memory read-after-write dependency.

## Validation After Fix
After splitting normalization into a second Triton kernel:
renormalize=True max reference error: approximately 4.47e-08
max row-sum error: approximately 2.38e-07
20 repeated executions are deterministic, with only negligible FP32 rounding noise
expert ids remain correct

In a vLLM Qwen3.6-35B-A3B TP=4, 64-concurrent decode reproduction, the original path produced around 21-24 corrupted requests per round. The fixed kernel reduces this to the same small residual level as independent PyTorch and AscendC router references.

## Attachments

[flaggems_topk_split_normalize.patch](https://github.com/user-attachments/files/30593079/flaggems_topk_split_normalize.patch)
[tmp_probe_flaggems_topk_renorm.py](https://github.com/user-attachments/files/30593081/tmp_probe_flaggems_topk_renorm.py)
[environment.txt](https://github.com/user-attachments/files/30593080/environment.txt)

**Comment by luentong** [comment]
I opened a PR ready for review with a minimal fix and Ascend NPU regression test:

[<upstream-pr-url>](https://github.com/flagos-ai/FlagGems/pull/5125)

The fix moves `renormalize=True` into a separate Triton kernel, avoiding the
same-kernel global-memory read-after-write that corrupts router weights on the
Ascend Triton path.

Validation on Ascend 910B with CANN 9.0:

- BF16 input shape `(64, 256)`, top-k = 8
- `renormalize=False` matches the FP32 reference and is deterministic
- `renormalize=True` matches the reference, row sums are approximately 1, and
  repeated calls are deterministic
- Added an automated regression test for both modes

I would appreciate feedback on whether maintainers prefer this generic
two-kernel fix or an Ascend-specific implementation.
